### PR TITLE
Record k8s service info

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

When a k8s charm is deployed, a k8s service is created.
We record the details of that service against the application in state, so that the addresses can be used when the charm asks for network info.

Commit 1 is the state artefacts.
Commit 2 is the api layers
Commit 3 is the caas unit provisioner worker

## QA steps

Run up a caas model, deploy a charm, check mongo to ensure service details are stored.

### CODE REVIEW

File: apiserver/params/params.go [hunk 1]
```diff
@@ -498,6 +498,20 @@ type ApplicationUnitParams struct {
...
 
+// UpdateApplicationServiceArgs holds the parameters for
+// updating application services.
+type UpdateApplicationServiceArgs struct {
+	Args []ApplicationServiceParams `json:"args"`
+}
+
+// ApplicationServiceParams holds parameters used to update
+// an application's service definition for the cloud.
+type ApplicationServiceParams struct {
```
Comment (babbageclunk) [inline apiserver/params/params.go]: The names for these types are pretty confusing, especially in the client code. `ApplicationServiceArg` would be clearer for this one, even though it's composed of multiple values.

File: caas/broker.go [hunk 2]
```diff
@@ -45,6 +49,12 @@ type Broker interface {
 	Units(appName string) ([]Unit, error)
 }
 
+// Service represents information about the status of a "service".
```
Comment (babbageclunk) [inline caas/broker.go]: Seems a bit sarcastic. :)

File: apiserver/params/params.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/params/params.go]: ok

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**